### PR TITLE
README: Reorganization and new "Using programmatically" section

### LINKED ISSUE #1719: README: add note that testing and results are done locally (no remote servers)
I searched the docs on this extension and couldn't find a clear answer.

I regularly want to use tools like lighthouse to test content loaded in the browser from a in internal (Intranet) site that contains high security information that should not ever leave my offices walls.

Are all of the tests run locally in the browser? or is some of it processed externally?

If this is all run locally in the browser, can the documentation be updated to indicate as such?

### PR DESCRIPTION
**Pull request by ebidel**
R: all

Also fixes #1719. 

The example is extracted and simplified from lighthouse hue lights thing I made. Happy to rework it a little, but it's pretty chill and is known to be everything you need to call LH programmatically.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -200,104 +107,224 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
...
+}
+```
+
+Then, run with: `lighthouse --config-path=config.json http://www.random.url`
 
 The traceviewer-based trace processor from [node-big-rig](https://github.com/GoogleChrome/node-big-rig/tree/master/lib) was forked into Lighthouse. Additionally, the [DevTools' Timeline Model](https://github.com/paulirish/devtools-timeline-model) is available as well. There may be advantages for using one model over another.
 
+### Trace processing
+
 **To update traceviewer source:**
```
Comment (paulirish) [inline readme.md]: this is dev instructions for us, and belongs in the contributing.md

File: readme.md [hunk 2]
```diff
@@ -200,104 +107,224 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
...
+node lighthouse-cli http://example.com
 ```
 
-## Code Style
+> **Getting started tip**: `node --inspect --debug-brk lighthouse-cli http://example.com` to open up Chrome DevTools and step
+through the entire app. See [Debugging Node.js with Chrome
+DevTools](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27#.59rma3ukm)
+for more info.
+
+## Coding Style
```
Comment (paulirish) [inline readme.md]: move to contributing.md

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# Architecture
+
+_Some incomplete notes_
+
+## Components
+
+* **Driver** - Interfaces with [Chrome Debugging Protocol](https://developer.chrome.com/devtools/docs/debugger-protocol)  ([API viewer](https://chromedevtools.github.io/debugger-protocol-viewer/))
+* **Gathers** - Requesting data from the browser (and maybe post-processing)
```
Comment (patrickhulce) [inline docs/architecture.md]: Gatherers?

File: readme.md [hunk 4]
```diff
@@ -200,104 +107,224 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
...
+```sh
+git clone https://github.com/GoogleChrome/lighthouse
+
+cd lighthouse
+npm install
+npm run install-all
+
+# The CLI is authored in TypeScript and requires compilation.
+# If you need to make changes to the CLI, run the TS compiler in watch mode:
+# cd lighthouse-cli && npm run dev
```
Comment (paulirish) [inline readme.md]: link to contributing and architecture somewhere around here?

File: readme.md [hunk 5]
```diff
@@ -311,8 +338,29 @@ git -C "./third_party/src/catapult/" pull
...
+web applications. By running a set of tests, developers can discover new web platform APIs, become
+aware of performance pitfalls, and learn (newer) best practices. In other words, do better on the web!
+
+DBW is implemented as a set of standalone [gatherers](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/gather/gatherers/dobetterweb) and [audits](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/audits/dobetterweb) that are run alongside the core Lighthouse tests. The tests show up under "Best Practices" in the report.
+
+If you'd like to contribute, check the [list of issues](https://github.com/GoogleChrome/lighthouse/issues?q=is%3Aissue+is%3Aopen+label%3ADoBetterWeb) or propose a new audit by filing an issue.
+
+### Are results sent to a remote server?
+
+Hope. Lighthouse runs locally, auditing a page using a local version of the Chrome browser installed the
```
Comment (paulirish) [inline readme.md]: Hope 🙏

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+* **Artifacts** - The output of gatherers
+* **Audits** - Non-performance evaluations of capabilities and issues. Includes a raw value and score of that value.
+* **Metrics** - Performance metrics summarizing the UX
+* **Diagnoses** - The perf problems that affect those metrics
```
Comment (patrickhulce) [inline docs/architecture.md]: Is this still a thing that we call Diagnoses?

File: readme.md [hunk 7]
```diff
@@ -311,8 +338,29 @@ git -C "./third_party/src/catapult/" pull
...
+machine. Report results are never processed or beaconed to a remote server.
+
 <p align="center">
-  <img src="https://cloud.githubusercontent.com/assets/39191/22478294/23f662f6-e79e-11e6-8de3-ffd7be7bf628.png" alt="Lighthouse logo" height="300">
+  <img src="https://cloud.githubusercontent.com/assets/39191/22478294/23f662f6-e79e-11e6-8de3-ffd7be7bf628.png" alt="Lighthouse logo" height="150">
   <br>
   <b>Lighthouse</b>, ˈlītˌhous (n): a <s>tower or other structure</s> tool containing a beacon light to warn or guide <s>ships</s> developers at "sea".
```
Comment (paulirish) [inline readme.md]: > <s>ships at sea</s> developers.

IMO slightly better. if you want.

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
+# Architecture
```
Comment (paulirish) [inline docs/architecture.md]: IMO this file can get merged with API-and-internals.md

and since readme and contributing are in root, i kinda prefer keeping this md file there rather than docs/. wdyt?

File: readme.md (hunk 4 above)
Comment (paulirish) [inline readme.md]: and can you link to architecture doc from contributing.md?

File: readme.md [hunk 9]
```diff
@@ -200,104 +107,224 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
 $ lighthouse --disable-device-emulation --disable-cpu-throttling https://mysite.com
 ```
 
-## Tests
+## Using programmatically
 
-Some basic unit tests forked are in `/test` and run via mocha. ESLint also checks for style violations.
+The example below shows how to setup and run Lighthouse programmatically as a Node module.
```
Comment (patrickhulce) [inline readme.md]: How important is supporting use as a node module to us? If we think it's important, any issues to make basic usage easier than 80 lines? If not, maybe let's move this to separate file and link to it from README.

Personally, I'd love to see chrome launcher published to NPM separately with a slightly friendlier interface (happy to take this up if there's agreement), and a more organized programmatic API that doesn't involve so much digging into our directory tree (more resilient to doing our rename of `lighthouse-cli -> cli` etc

File: readme.md (hunk 4 above)
Comment (ebidel) [inline readme.md]: done

File: readme.md (hunk 2 above)
Comment (ebidel) [inline readme.md]: done

File: readme.md (hunk 1 above)
Comment (ebidel) [inline readme.md]: done

File: readme.md (hunk 5 above)
Comment (ebidel) [inline readme.md]: ha

File: readme.md (hunk 7 above)
Comment (ebidel) [inline readme.md]: done

File: docs/architecture.md (hunk 8 above)
Comment (ebidel) [inline docs/architecture.md]: Up to you. I figured we'd have more and wanted to declutter.

File: docs/architecture.md (hunk 3 above)
Comment (ebidel) [inline docs/architecture.md]: done

File: docs/architecture.md (hunk 6 above)
Comment (ebidel) [inline docs/architecture.md]: removed

File: readme.md (hunk 9 above)
Comment (ebidel) [inline readme.md]: Open sourcing a more convenient launcher would be great.  That's really the bulk of the example, and something I struggled to get working for my demo.

When looking into a lib, I do really like seeing install + different usages in the main readme. Since a lot of people as for docs on this, I'd love to keep it front and center. Ironically, and email thread from Addy prompted me to add this...which led to a full reorg :)

Happy to move it out though.

Comment (patrickhulce) [inline readme.md]: > When looking into a lib, I do really like seeing install + different usages in the main readme.

Totally agree, I just also would find it off-putting if the usage example was so complicated :) I say let's leave it here, and I'll file issue to improve the API.

File: docs/architecture.md (hunk 8 above)
Comment (ebidel) [inline docs/architecture.md]: There's also https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/#publishing-your-github-pages-site-from-a-docs-folder-on-your-master-branch, which we could leverage someday.

Comment (ebidel) [inline docs/architecture.md]: Moved and renamed that file to "headless-chrome.md". It didn't contain anything but headless Chrome stuff :)

File: readme.md [hunk 10]
```diff
@@ -200,119 +107,244 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
...
+const lighthouse = require('lighthouse');
+const ChromeLauncher = require('lighthouse/lighthouse-cli/chrome-launcher.js').ChromeLauncher;
+const Printer = require('lighthouse/lighthouse-cli/printer');
+const yargs = require('yargs');
+
+// Also could use your own custom config.
+const PERF_CONFIG = require('lighthouse/lighthouse-core/config/perf.json');
+// const DEFAULT_CONFIG = require('lighthouse/lighthouse-core/config/default.json');
+
+const _SIGINT = 'SIGINT';
```
Comment (brendankenny) [inline readme.md]: Doesn't this seem like overkill? Handling command line args and SIGINT seems out of scope for a readme example

File: readme.md [hunk 11]
```diff
@@ -200,119 +107,244 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
...
+      .catch(err => {
+        // Kill Chrome if there's an error.
+        return this.launcher.kill().then(() => {
+          throw err;
+        }, handleError);
+      });
+  }
+
+  // Example of extracting the overall score from the results.
+  getOverallScore(results) {
```
Comment (brendankenny) [inline readme.md]: Ha, seems like we should already be including this in the lighthouse results

**Comment by ebidel** [comment]
PTAL

File: readme.md [hunk 12]
```diff
@@ -200,119 +107,244 @@ $ adb forward tcp:9222 localabstract:chrome_devtools_remote
...
+    const scoredAggregations = results.aggregations.filter(a => a.scored);
+    const total = scoredAggregations.reduce((sum, aggregation) => sum + aggregation.total, 0);
+    return (total / scoredAggregations.length) * 100;
+  }
+}
+
+const flags = yargs
```
Comment (brendankenny) [inline readme.md]: yargs especially seems like it could be cut. 9 times out of 10 it seems like people will be calling lighthouse within an existing script or with hardcoded values

File: docs/architecture.md (hunk 8 above)
Comment (paulirish) [inline docs/architecture.md]: sg

File: readme.md (hunk 12 above)
Comment (paulirish) [inline readme.md]: +1 to this. 

you can see https://github.com/paulirish/pwmetrics/blob/27ee73fd1c9c8ed5457eebafb245b1592ac26a38/lib/index.js#L64-L90 which is how we adopt LH in pwmetrics. it's pretty small.

File: readme.md (hunk 11 above)
Comment (ebidel) [inline readme.md]: Sure, but I don't think it hurts to show an "Example of extracting" things from our crazy config object.

File: readme.md (hunk 12 above)
Comment (ebidel) [inline readme.md]: done

File: readme.md (hunk 11 above)
Comment (ebidel) [inline readme.md]: kept, but moved it as a separate function example

**Comment by ebidel** [comment]
Removed the extras from the example

**Comment by paulirish** [comment]
Yeah this small example is clutch. Nice!

**Comment by ebidel** [comment]
@patrickhulce how does this look?

File: readme.md (hunk 11 above)
Comment (brendankenny) [inline readme.md]: ha, sorry, I meant it's ridiculous that the results object don't currently include the overall score and that it has to be extracted :) Somehow I hadn't noticed that in all this time. We should add it.

Comment (ebidel) [inline readme.md]: Gotcha. I was surprised too. But I guess we've also never come across it. We have the notion of scores for each aggregation, but only have 1 that's scored atm.
